src/Lights/Shadows/shadowGenerator.ts

```diff
@@ -656,10 +681,100 @@ export class ShadowGenerator implements IShadowGenerator {
         if (this._shadowMap2) {
             return this._shadowMap2;
         }
-
         return this._shadowMap;
     }
 
+    private _viewSpaceFrustums: Array<Array<Vector3>>;
+    private _viewSpaceBoundingSpheres: Array<BoundingSphere>;
+    private _CSMtransformMatrices: Array<Matrix>;
+    private _cascades = 1;
+    /**
+     * Gets the number of cascades used in CSM mode.
+     */
+    public get cascades(): number {
+        return this._cascades;
+    }
+    /**
+     * Sets the number of cacades used in CSM mode.
+	 * Limited to 2 to 4 cascade to reduce resource footprint
+     */
+    public set cascades(value: number) {
+        if (this._cascades === value || value < ShadowGenerator.MIN_CASCADES || value > ShadowGenerator.MAX_CASCADES) {
+            return;
+        }
+        this._cascades = value;
+        if (this._cascades >= ShadowGenerator.MIN_CASCADES && this._cascades <= ShadowGenerator.MAX_CASCADES) {
+            this.useCSM = true;
+        }
+    }
+
+    private _useCSM = false;
+    /**
+     * Gets the CSM mode.
+     */
+    public get useCSM(): boolean {
+        return this._useCSM;
+    }
+    /**
+     * Sets the CSM mode.
+     */
+    public set useCSM(value: boolean) {
+        if (this._useCSM === value) {
+            return;
+        }
+        this._useCSM = value;
+    }
+
+    private _csmFrustumLength: number;
+    /**
+     * Gets the csmFrustumLength value: furthest range of the frustum for the CSM mode.
+     */
+    public get csmFrustumLength(): number {

```

Comment:
Why not relying on the camera or light minShadowZ  and MaxShadowZ like the rest of the shadows ?

Reply:
afaik, they mean different things.
a) camera.maxZ vs csmFrustumLength
Using racing game as example, scenery from driver seat falls within camera minZ and maxZ limits. So a far building say, 1000 units away does not need to be in CSM range but the competing car in the next lane should be in csmFrustumLength range. This allows the user to configure its scene appropriately. If user wishes to set csmFrustumLength = camera.maxZ, that is still fine but if the user does not wish to, he/she must have the option not to.

b) light.shadowMaxZ vs csmFrustumLength
How far a light should cast its shadows and the csm frustum length are independent from each other. Imagine camera, a sphere 100 units away, and a light behind the sphere 5 units away all in a straight line: camera ---100--- sphere -5- light. If light..shadowMaxZ is 20, then the camera should see the sphere's shadow, right ? But if the csmFrustumLength is also 20, the camera will see a sphere with no shadow.